# Student ID: XXX

**You student_id is your 7/8 digit faser number.**

This is a sample format for CE807-SU: Assignment 2. You must follow the format.
The code will have three broad sections, and additional section, if needed,


1.   Common Codes
2.   Supervised Training Codes
3.   Other Method/model Codes, if any

**You must have `train_method` This will be evaluated automatically, without this your code will fail and no marked.**

You code should be proverly indended, print as much as possible, follow standard coding (https://peps.python.org/pep-0008/) and documentaion (https://colab.research.google.com/github/jakevdp/PythonDataScienceHandbook/blob/master/notebooks/01.01-Help-And-Documentation.ipynb) practices.

Before each `code cell`, you must have a `text cell` which explain what code cell is going to do. For each function/class, you need to properly document what are it's input, functionality and output.

If you are using any non-standard library, you must have command to install that, for example `pip install datasets`.

You must print `train`, `validation` and `test` performance measures.

You must also print `train` and `validation` loss in each `epoch`, wherever you are using `epoch`, say in any deep learning algorithms.

Your code must

*   To reproducibality of the results you must use a `seed`, you have to set seed in `torch`, `numpy` etc, use same seed everywhere **and your Student ID should be your seed**.
*   read dataset from './CE807-SU/Assignment2/student_id/' folder which will have 3 files [`train.csv`, `val.csv`, `test.csv`]
*   save model after finishing the training in './CE807-SU/Assignment2/student_id/models/XXX/' where XXX = [1,2]
*   at testing time you will load models from './CE807-SU/Assignment2/student_id/models/XXX/' where XXX = [1,2] and then test on your data, and save the output in the same folder

*   **Your output file based on the test file will be named `output_test.csv` and will have the existing columns from test.csv and `out_label`** You need to save file in the respective model folders.




**Install and import all required libraries first before starting to code.**








Let's install all require libraries. For example, `transformers`

Let's import all require libraries.
For example, `numpy`

**Let's put your student id as a variable, that you will use different places**

In [24]:
student_id = 2214000 # Note this is an interger and you need to input your id

Let's set `seed` for all libraries like `torch`, `numpy` etc as my student id

In [25]:
# set same seeds for all libraries
import numpy as np
#numpy seed
np.random.seed(student_id)

In [26]:
!pip install transformers

In [27]:
import torch
print(torch.__version__)

2.0.1+cu118


In [28]:
import torch
import torch.nn as nn

# Import pprint, module we use for making our print statements prettier
import pprint
pp = pprint.PrettyPrinter()

## Setup Code
Before getting started we need to run some boilerplate code to set up our environment. You'll need to rerun this setup code each time you start the notebook.

First, run this cell load the [autoreload](https://ipython.readthedocs.io/en/stable/config/extensions/autoreload.html?highlight=autoreload) extension. This allows us to edit `.py` source files, and re-import them into the notebook for a seamless editing and debugging experience.

In [29]:
import numpy as np
import os
import pandas as pd

In [30]:
%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


**Let's first allow the GDrive access and set data and model paths**

For examples,

student_id = 1234567

set GOOGLE_DRIVE_PATH_AFTER_MYDRIVE = ‘./CE807-SU/Assignment2/student_id/’ in your GDrive

now set all global variable,


train_file = os.path.join(GOOGLE_DRIVE_PATH_AFTER_MYDRIVE, 'train.csv')

MODEL_1_DIRECTORY = os.path.join('gdrive', 'MyDrive', GOOGLE_DRIVE_PATH, ‘models’, ‘1’)


model_1_output_test_file = os.path.join(MODEL_1_DIRECTORY,'output_test.csv')

Sample output directory and file structure: https://drive.google.com/drive/folders/1ohkXpVvXCl3vJpWGhSlgjqjsno27O5KP?usp=sharing

In [31]:
# Mount Google Drive
from google.colab import drive
drive.mount('/content/gdrive', force_remount=True)

Mounted at /content/gdrive


In [32]:
import sys


In [33]:
# Add your code to initialize GDrive and data and models paths

# TODO: Fill in the Google Drive path where you uploaded the assignment, data and code
# Example: If your student_id is 1234567 then your directory will be './CE807-SU/Assignment2/1234567/'

GOOGLE_DRIVE_PATH_AFTER_MYDRIVE = os.path.join('./CE807-SU/Assignment2/',str(student_id)) # Make sure to update with your student_id and student_id is an integer
GOOGLE_DRIVE_PATH = os.path.join('gdrive', 'MyDrive', GOOGLE_DRIVE_PATH_AFTER_MYDRIVE)
sys.path.append(GOOGLE_DRIVE_PATH)
print('List files: ', os.listdir(GOOGLE_DRIVE_PATH))

test_file = os.path.join(GOOGLE_DRIVE_PATH, 'test.csv')
print('Test file: ', test_file)

MODEL_1_DIRECTORY = os.path.join(GOOGLE_DRIVE_PATH, 'models', '1') # Model 2 directory
print('Model 1 directory: ', MODEL_1_DIRECTORY)



List files:  ['testSup.ipynb', 'test.csv', 'train.csv', 'valid.csv', '__pycache__', 'models', 'trainSup.ipynb', 'testUnSup.ipynb', 'common_code.py', 'trainUnSup.ipynb']
Test file:  gdrive/MyDrive/./CE807-SU/Assignment2/2214000/test.csv
Model 1 directory:  gdrive/MyDrive/./CE807-SU/Assignment2/2214000/models/1


# Common Codes

You will write all common codes in `common_code.py` file and use train/test for both models, for examples


*   Data read
*   Performance Matrics
*   Print Dataset Statistics
*   Saving model and output (during training)
*   Loading Model and output (during testing)
*   etc




In [34]:
from common_code import read_csv_dataset,prepare_dataset_nn,prepare_dataset,calculate_dataset_details,compute_performance,load_model,text_preprocessing,save_model,create_feature_matrix,lemmatized_text,preprocess_and_topic_model,get_dominant_topic,save_label_encoder,save_lda_model_and_dictionary,save_model_hypertuning,balance_dataset,collate_batch,HateSpeechClassificationMLP,HateSpeechClassificationMLPDataset,save_model_nn,predict,augment_data,prepare_dataset,save_model_machine_learning,predict,load_label_encoder

**Let's see test file**

In [35]:
test_df = read_csv_dataset(test_file)

Read  gdrive/MyDrive/./CE807-SU/Assignment2/2214000/test.csv
      id                                              tweet label
0   A272  @USER  question for our home game against @USE...   NOT
1  A2244  i dont get it, why does roblox shit on you for...   OFF
2  A1564  The left blames President Trump for everything...   OFF
3  A2081                      @USER thanks for your follow!   NOT
4  A3306            An artist will fall in love with a fan.   NOT


**The below code checks if a CUDA-enabled GPU (NVIDIA GPU) is available for PyTorch computations.**

In [36]:
import torch

if torch.cuda.is_available():
  device = torch.device("cuda")
else:
  device = torch.device("cpu")
device

device(type='cpu')

## Text Pre-processing

In [37]:
import nltk
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [38]:
# Apply text preprocessing to the 'text' column
test_df['preprocessed_text'] = test_df['tweet'].apply(text_preprocessing)

# Display the preprocessed data
print(test_df.head())

      id                                              tweet label  \
0   A272  @USER  question for our home game against @USE...   NOT   
1  A2244  i dont get it, why does roblox shit on you for...   OFF   
2  A1564  The left blames President Trump for everything...   OFF   
3  A2081                      @USER thanks for your follow!   NOT   
4  A3306            An artist will fall in love with a fan.   NOT   

                                   preprocessed_text  
0   user question home game  user buy ticket even...  
1                  dont get  roblox shit using rthro  
2  left blames president trump everything  gives ...  
3                                user thanks follow   
4                              artist fall love fan   


In [39]:
# Access the 'label' column
label = test_df['label'].to_numpy()

unique, counts = np.unique(label, return_counts=True)

print("Total number of tweets: " + str(label.size))
for i in range(len(unique)):
    print("Class " + str(unique[i]) + " has " + str(counts[i]) + " tweets.")

Total number of tweets: 800
Class NOT has 578 tweets.
Class OFF has 222 tweets.


In [40]:
from torch.utils.data import Dataset, DataLoader, TensorDataset

In [41]:
# Mapping of label values to numeric values
label_mapping = {'NOT': 0, 'OFF': 1}

# Convert 'label' column to numeric form
test_df['label'] = test_df['label'].map(label_mapping)

#Supervised Method Start (Method 1)

In this section you will write all details of your Method 1.

You will have to enter multiple `code` and `text` cell.

Your code should follow the standard ML pipeline


*   Data reading
*   Data clearning, if any
*   Convert data to vector/tokenization/vectorization
*   Model Declaration/Initialization/building
*   Training and validation of the model using training and validation dataset
*   Save the trained model
*   Load and Test the model on testing set
*   Save the output of the model


You could add any other step(s) based on your method's requirement.

After finishing the above, you need to usd splited data as defined in the assignment and then do the same for all 4 sets. Your code should not be copy-pasted 4 time, make use of `function`.


## Testing Method Code
Your test code should be a stand alone code that must take `test_file`, `model_file` and `output_dir` as input. You could have other things as also input, but these three are must. You would load both files, and generate output based on inputs. Then you will `print` / `display`/ `plot` all performance metrics, and save the output file in the `output_dir`

**The below test_method function tests a trained neural network model on a dataset, loads the model, performs predictions, computes evaluation metrics, and saves output to a specified directory.**

In [42]:
from sklearn.metrics import confusion_matrix
import pickle

def test_method(test_file, model_file, vectorizer_file, output_dir):
    """
     take test_file, model_file and output_dir as input.
     It loads model and test of the examples in the test_file.
     It prints different evaluation metrics, and saves the output in output directory

     ADD Other arguments, if needed

    Args:
        test_file: test file name
        model_file: model file name
        output_dir: Output Directory
    """
    reverse_label_mapping = {0: 'NOT', 1: 'OFF'}
    test_label = test_df['label']

    model_mlp, tokenizer = load_model(model_file, token_file)

    # Load the vocabulary from disk
    vocab_file = MODEL_1_DIRECTORY + "/vocab.sav"
    with open(vocab_file, "rb") as file:
        vocab = pickle.load(file)

    # Create your custom dataset and DataLoader for test set
    test_datasets = HateSpeechClassificationMLPDataset(test_df, tokenizer, vocab)

    test_loaders = DataLoader(test_datasets, batch_size=32, collate_fn=collate_batch)

    test_gt_labels, test_pred_labels = predict(model=model_mlp,device=device, dataloader=test_loaders)
    # Compute F1 score and print confusion matrix for the test set
    test_f1_score = compute_performance(test_gt_labels, test_pred_labels, split='test')
    print("Confusion matrix - Test set")
    print(confusion_matrix(test_gt_labels, test_pred_labels))

    # Add the predicted labels to the test_df DataFrame
    test_df['label'] = test_gt_labels
    test_df['label'] = test_df['label'].map(reverse_label_mapping)
    test_df['out_label'] = test_pred_labels
    test_df['out_label'] = test_df['out_label'].map(reverse_label_mapping)

    out_file = os.path.join(output_dir, 'output_test.csv')

    print('Saving model output to', out_file)
    test_df.to_csv(out_file)


In [43]:
model_file  = MODEL_1_DIRECTORY + '/model.sav'
token_file = MODEL_1_DIRECTORY + '/tokenizer.sav'

test_method(test_file, model_file, token_file, MODEL_1_DIRECTORY)

Loaded model from gdrive/MyDrive/./CE807-SU/Assignment2/2214000/models/1/model.sav
Loaded Vectorizer from gdrive/MyDrive/./CE807-SU/Assignment2/2214000/models/1/tokenizer.sav
Computing different preformance metrics on test  set of Dataset
F1 Score(macro):  0.5728330636016338
Accuracy:  0.7575
Confusion matrix - Test set
[[566  12]
 [182  40]]
Saving model output to gdrive/MyDrive/./CE807-SU/Assignment2/2214000/models/1/output_test.csv


**The test set performance metrics show a lower F1 score (macro) of 0.5728 compared to the training and validation sets, indicating that the model's performance on the test data is not as strong as on the training data. The accuracy is 75.5%, but it should be considered in conjunction with other performance metrics. The confusion matrix provides insights into the model's performance, with true positives (TP) at 40, true negatives (TN) at 566, false positives (FP) at 12, and false negatives (FN) at 182. The model seems better at correctly classifying true negative samples (class 0) than true positive samples (class 1), but struggles to classify true positive samples correctly due to the high number of false negatives (FN = 189).**

#Supervised Method End (Method 1)

# Other Method/model Start

##Other Method/model End

In [44]:
file1 = '/content/gdrive/MyDrive/CE807-SU/Assignment2/2214000/models/2/output_test.csv'
df1 = pd.read_csv(file1)

# Load the second CSV file
file2 = '/content/gdrive/MyDrive/CE807-SU/Assignment2/2214000/models/1/output_test.csv'
df2 = pd.read_csv(file2)

# Merge the two DataFrames on the common "preprocessed_text" column
common_records = pd.merge(df1, df2, on=['preprocessed_text'])
common_records

Unnamed: 0_x   id_x                                            tweet_x  \
0               0   A272  @USER  question for our home game against @USE...   
1               1  A2244  i dont get it, why does roblox shit on you for...   
2               2  A1564  The left blames President Trump for everything...   
3               3  A2081                      @USER thanks for your follow!   
4               4  A3306            An artist will fall in love with a fan.   
..            ...    ...                                                ...   
795           795   A319  @USER @USER @USER @USER @USER @USER orisa ball...   
796           796   A906  @USER What is the alternative ?? What is their...   
797           797  A1877  Licking pussy is for kids, men eat it.   Serve...   
798           798  A1726  My favorite part of the day is when I get to g...   
799           799  A3654                                 @USER @USER On it!   

    label_x                                  preprocessed_text  \
0       NOT   user question home game  user buy ticket even...   
1       OFF                  dont get  roblox shit using rthro   
2       OFF  left blames president trump everything  gives ...   
3       NOT                                user thanks follow    
4       NOT                              artist fall love fan    
..      ...                                                ...   
795     NOT   user  user  user  user  user  user orisa ball...   
796     NOT   user alternative   offer   destroyed parks to...   
797     OFF        licking pussy kids  men eat  serve dinner 🍽   
798     NOT                      favorite part day get go bed    
799     NOT                                        user  user    

     dominant_topic  bmu_x  bmu_y  topic_cluster out_label_x  Unnamed: 0_y  \
0                 5      2      5              0         NOT             0   
1                 1      0      7              0         NOT             1   
2                 2      1      6              0         NOT             2   
3                 7      5      0              0         NOT             3   
4                 5      2      5              0         NOT             4   
..              ...    ...    ...            ...         ...           ...   
795               2      1      6              0         NOT           795   
796               7      5      0              0         NOT           796   
797               0      0      5              0         NOT           797   
798               2      0      6              0         NOT           798   
799               4      3      1              1         OFF           799   

      id_y                                            tweet_y label_y  \
0     A272  @USER  question for our home game against @USE...     NOT   
1    A2244  i dont get it, why does roblox shit on you for...     OFF   
2    A1564  The left blames President Trump for everything...     OFF   
3    A2081                      @USER thanks for your follow!     NOT   
4    A3306            An artist will fall in love with a fan.     NOT   
..     ...                                                ...     ...   
795   A319  @USER @USER @USER @USER @USER @USER orisa ball...     NOT   
796   A906  @USER What is the alternative ?? What is their...     NOT   
797  A1877  Licking pussy is for kids, men eat it.   Serve...     OFF   
798  A1726  My favorite part of the day is when I get to g...     NOT   
799  A3654                                 @USER @USER On it!     NOT   

    out_label_y  
0           NOT  
1           OFF  
2           NOT  
3           NOT  
4           NOT  
..          ...  
795         NOT  
796         NOT  
797         NOT  
798         NOT  
799         NOT  

[800 rows x 15 columns]

In [45]:
output_data = common_records[['label_x','preprocessed_text', 'out_label_x', 'out_label_y']]

# Rename the columns to match the desired output
output_data.columns = ['label', 'preprocessed_text', 'out_label_model1', 'out_label_model2']

# Save the output data to a new CSV file
output_file = 'common_records.csv'
output_data.to_csv(output_file, index=False)

print("Common records have been saved to", output_file)


Common records have been saved to common_records.csv


In [ ]:
file = '/content/common_records.csv'
ot = pd.read_csv(file)
